File: scripts/run_all_professional_systems_workflows.py
#!/usr/bin/env python3
"""Run professional systems workflows across all existing article folders."""

from pathlib import Path
import subprocess
import sys

repo = Path(__file__).resolve().parents[1]
articles = repo / "articles"
workflow_name = "run_professional_systems_workflow.py"

if not articles.exists():
    raise SystemExit(f"Missing articles directory: {articles}")

article_dirs = sorted([p for p in articles.iterdir() if p.is_dir()])
if not article_dirs:
    raise SystemExit("No article directories found.")

failures = []
for article_dir in article_dirs:
    workflow = article_dir / "python" / workflow_name
    if not workflow.exists():
        failures.append((article_dir.name, "missing workflow"))
        continue
    print(f"\n=== Python workflow: {article_dir.name} ===")
    result = subprocess.run([sys.executable, str(workflow)], cwd=str(article_dir))
    if result.returncode != 0:
        failures.append((article_dir.name, f"exit {result.returncode}"))

if failures:
    print("\nFailures:")
    for slug, detail in failures:
        print(f"- {slug}: {detail}")
    raise SystemExit(1)

print("\nAll professional Python workflows completed successfully.")

File: scripts/run_all_professional_systems_workflows.R
#!/usr/bin/env Rscript
# Run professional R workflows across all existing article folders.

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]), mustWork = FALSE)

if (length(script_path) == 0 || is.na(script_path) || script_path == "") {
  repo <- getwd()
} else {
  repo <- dirname(dirname(script_path))
}

articles_dir <- file.path(repo, "articles")
article_dirs <- list.dirs(articles_dir, full.names = TRUE, recursive = FALSE)

if (length(article_dirs) == 0) {
  stop("No article directories found.")
}

failures <- c()

for (article_dir in article_dirs) {
  workflow <- file.path(article_dir, "r", "run_professional_systems_workflow.R")
  slug <- basename(article_dir)
  if (!file.exists(workflow)) {
    failures <- c(failures, paste(slug, "missing workflow"))
    next
  }
  message("\n=== R workflow: ", slug, " ===")
  status <- system2("Rscript", workflow)
  if (!identical(status, 0L)) {
    failures <- c(failures, paste(slug, "exit", status))
  }
}

if (length(failures) > 0) {
  message("\nFailures:")
  for (failure in failures) {
    message("- ", failure)
  }
  quit(status = 1)
}

message("\nAll professional R workflows completed successfully.")
